execution-apis: Add EIP-7594 (PeerDAS) related changes (#630)
ethereum/execution-apis: src/engine/osaka.md

## Structures

### BlobsBundleV2

The fields are encoded as follows:

- `blobs`: `Array of DATA` - Array of blobs, each blob is `FIELD_ELEMENTS_PER_BLOB * BYTES_PER_FIELD_ELEMENT = 4096 * 32 = 131072` bytes (`DATA`) representing a SSZ-encoded `Blob` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844)
- `commitments`: `Array of DATA` - Array of `KZGCommitment` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).
- `cell_proofs`: `Array of DATA` - Array of `KZGProof` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).

`blobs` and `commitments` arrays **MUST** be of same length, `cell_proofs` contains exactly `CELLS_PER_EXT_BLOB` * `len(blobs)` cell proofs.

### BlobAndProofV2

The fields are encoded as follows:

- `blob`: `DATA` - `FIELD_ELEMENTS_PER_BLOB * BYTES_PER_FIELD_ELEMENT = 4096 * 32 = 131072` bytes (`DATA`) representing a SSZ-encoded `Blob` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844).
- `cell_proofs`: `Array of DATA` - Array of `KZGProof` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).

> `cell_proofs` contains exactly `CELLS_PER_EXT_BLOB` * `len(blobs)` cell proofs.

Reviewer: ```suggestion
`cell_proofs` contains exactly `CELLS_PER_EXT_BLOB` cell proofs.
```

  - `blockValue` : `QUANTITY`, 256 Bits - The expected value to be received by the `feeRecipient` in wei
  - `blobsBundle`: [`BlobsBundleV2`](#BlobsBundleV2) - Bundle with data corresponding to blob transactions included into `executionPayload`
  - `shouldOverrideBuilder` : `BOOLEAN` - Suggestion from the execution layer to use this `executionPayload` instead of an externally provided one
  - `executionRequests`: `Array of DATA` - Execution layer triggered requests obtained from the `executionPayload` transaction execution.
* error: code and message set in case an exception happens while getting the payload.

#### Specification

This method follows the same specification as [`engine_getPayloadV4`](./prague.md#engine_getpayloadv4) with changes of the following:

1. The call **MUST** return `BlobsBundleV2` with empty `blobs`, `commitments` and `cell_proofs` if the payload doesn't contain any blob transactions.

2. The call **MUST** return `blobs` and `cell_proofs` that match the `commitments` list, i.e. 
   1. `assert len(blobsBundle.commitments) == len(blobsBundle.blobs)` and
   2. `assert len(blobsBundle.cell_proofs) == len(blobsBundle.blobs) * CELLS_PER_EXT_BLOB` and

>    3. `assert verify_cell_kzg_proof_batch(blobsBundle.commitments, cell_indices, cells, blobsBundle.cell_proofs)`.

Author: @fradamt curious where should we get the indicies, or is it like a generalized array of [0, 1, ...., total number of cells]?

Reviewer: `cell_indices` should be `[0, ..., CELLS_PER_EXT_BLOB, 0, ..., CELLS_PER_EXT_BLOB, ...]`. In python `list(range(CELLS_PER_EXT_BLOB)) * len(blobsBundle.blobs)`. 

Also the list of commitments should be `[blobsBundle.commitments[i] for i in range(len(blobsBundle.blobs)) for _ in range(CELLS_PER_EXT_BLOB)] ` (repeating each commitment for each cell index. All of the inputs to `verify_cell_kzg_proof_batch` have the same length, `CELLS_PER_EXT_BLOB * num_blobs`)

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Structures](#structures)
  - [BlobsBundleV2](#blobsbundlev2)
  - [BlobAndProofV2](#blobandproofv2)
- [Methods](#methods)
  - [engine_getPayloadV5](#engine_getpayloadv5)
    - [Request](#request)
    - [Response](#response)
    - [Specification](#specification)
  - [engine_getBlobsV2](#engine_getblobsv2)
    - [Request](#request-1)
    - [Response](#response-1)
    - [Specification](#specification-1)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Structures

### BlobsBundleV2

The fields are encoded as follows:

- `blobs`: `Array of DATA` - Array of blobs, each blob is `FIELD_ELEMENTS_PER_BLOB * BYTES_PER_FIELD_ELEMENT = 4096 * 32 = 131072` bytes (`DATA`) representing a SSZ-encoded `Blob` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844)
- `commitments`: `Array of DATA` - Array of `KZGCommitment` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).

> - `cell_proofs`: `Array of DATA` - Array of `KZGProof` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).

Reviewer 2: ```suggestion
- `cell_proofs`: `Array of DATA` - Array of `KZGProof` (48 bytes each, type defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), semantics defined in [EIP-7594](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7594.md)).
```

Suggesting a slight modification for clarity: the type `KZGProof` is indeed defined in EIP-4844, but its semantics (i.e., what data it proves) are defined in EIP-7594. This minor tweak might help avoid confusion.

# Engine API -- Osaka

Engine API changes introduced in Osaka.

This specification is based on and extends [Engine API - Prague](./prague.md) specification.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Structures](#structures)
  - [BlobsBundleV2](#blobsbundlev2)
  - [BlobAndProofV2](#blobandproofv2)
- [Methods](#methods)
  - [engine_getPayloadV5](#engine_getpayloadv5)
    - [Request](#request)
    - [Response](#response)
    - [Specification](#specification)
  - [engine_getBlobsV2](#engine_getblobsv2)
    - [Request](#request-1)
    - [Response](#response-1)
    - [Specification](#specification-1)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Structures

### BlobsBundleV2

The fields are encoded as follows:

- `commitments`: `Array of DATA` - Array of `KZGCommitment` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).

> - `cell_proofs`: `Array of DATA` - Array of `KZGProof` (48 bytes each, type defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), semantics defined in [EIP-7594](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7594.md)).

Reviewer 3: this shouldn't be snake case imo, everything in the engine api uses camel case

Author: Will do along other potential changes later

- `blobs`: `Array of DATA` - Array of blobs, each blob is `FIELD_ELEMENTS_PER_BLOB * BYTES_PER_FIELD_ELEMENT = 4096 * 32 = 131072` bytes (`DATA`) representing a SSZ-encoded `Blob` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844)

> `blobs` and `commitments` arrays **MUST** be of same length, `proofs` contains exactly `CELLS_PER_EXT_BLOB` * `len(blobs)` cell proofs.

Reviewer 4: The phrasing here is a little awkward. I would suggest changing it to be:
```suggestion
`blobs` and `commitments` arrays **MUST** be of same length and `proofs` **MUST** contain exactly `CELLS_PER_EXT_BLOB` * `len(blobs)` cell proofs.
```

### BlobAndProofV2

The fields are encoded as follows:

- `blob`: `DATA` - `FIELD_ELEMENTS_PER_BLOB * BYTES_PER_FIELD_ELEMENT = 4096 * 32 = 131072` bytes (`DATA`) representing a SSZ-encoded `Blob` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844).
- `proofs`: `Array of DATA` - Array of `KZGProof` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).

> `proofs` contains exactly `CELLS_PER_EXT_BLOB` cell proofs.

Reviewer 4: Similarly, for consistency, I might say this:
```suggestion
`proofs` **MUST** contain exactly `CELLS_PER_EXT_BLOB` cell proofs.
```

1. The call **MUST** return `BlobsBundleV2` with empty `blobs`, `commitments` and `proofs` if the payload doesn't contain any blob transactions.

2. The call **MUST** return `blobs` and `proofs` that match the `commitments` list, i.e. 
   1. `assert len(blobsBundle.commitments) == len(blobsBundle.blobs)` and
   2. `assert len(blobsBundle.proofs) == len(blobsBundle.blobs) * CELLS_PER_EXT_BLOB` and
   3. `assert verify_cell_kzg_proof_batch(commitments, cell_indices, cells, blobsBundle.proofs)` (see [EIP-7594 consensus-specs](https://github.com/ethereum/consensus-specs/blob/36d80adb44c21c66379c6207a9578f9b1dcc8a2d/specs/fulu/polynomial-commitments-sampling.md#verify_cell_kzg_proof_batch))
      1. `cell_indices` should be `[0, ..., CELLS_PER_EXT_BLOB, 0, ..., CELLS_PER_EXT_BLOB, ...]`. In python, `list(range(CELLS_PER_EXT_BLOB)) * len(blobsBundle.blobs)`
      2. `commitments` should list each commitment `CELLS_PER_EXT_BLOB` times, repeating it for every cell. In python, `[blobsBundle.commitments[i] for i in range(len(blobsBundle.blobs)) for _ in range(CELLS_PER_EXT_BLOB)]`

>       3. All of the inputs to `verify_cell_kzg_proof_batch` have the same length, `CELLS_PER_EXT_BLOB * len(blobsBundle.blobs)`

Reviewer 4: Two things.

1. We should define what these are in the same order as the function parameters. So commitments, indices, cells, then proofs. Currently, we mention indices before commitments.
2. We don't tell the reader how to compute `cells`. I wouldn't consider this to be obvious. Something like:

* `cells` is the list of cells for an extended blob. In python, `[compute_cells(blob) for blob in blobsBundle.blobs]` (see [`compute_cells`](https://github.com/ethereum/consensus-specs/blob/v1.5.0-beta.3/specs/fulu/polynomial-commitments-sampling.md#compute_cells) in consensus-specs)

### engine_getBlobsV2

Consensus layer clients **MAY** use this method to fetch blobs from the execution layer blob pool.

> *Note*: This is a new optional method introduced after Pectra.

Reviewer 4: This note doesn't really make sense anymore. We can delete it. The previous `getBlobsV1` note existed because we added the endpoint after Cancun.

#### Request

> * method: `engine_getBlobsV2`

Reviewer 5: i am not sure what the relevance of this method is in peerdas unless this is meant to provide support for distributed block building https://notes.ethereum.org/@dankrad/BkJMU8d0R in which case the request/response format will fully change

Reviewer: Why would it change?

Reviewer: Those are ideas for the future, we are not going to be sharding the mempool in Fusaka.

Reviewer 5: right, then this will be useful only if we get all blobs from the EL if my understanding is correct?

Author: That is correct, and it could be used to do distributed blob publishing, happen on nodes who have all the blobs available locally

* params:
  1. `Array of DATA`, 32 Bytes - Array of blob versioned hashes.
* timeout: 1s

#### Response

* result: `Array of BlobAndProofV2` - Array of [`BlobAndProofV2`](#BlobAndProofV2), items of which may be `null`.
* error: code and message set in case an error occurs during processing of the request.

#### Specification

Refer to the specification for [`engine_getBlobsV1`](./cancun.md#engine_getblobsv1) with changes of the following:

> 1. return `BlobAndProofV2` instead of `BlobAndProofV1`

Reviewer 4: I think these specifications are a bit light. I'm not 100% sure, but I think this should use `engine_getBlobsV1` as a starting point. Also, this sentence is missing punctuation.

https://github.com/ethereum/execution-apis/blob/40088597b8b4f48c45184da002e27ffc3c37641f/src/engine/cancun.md?plain=1#L204-L214

# Engine API -- Osaka

Engine API changes introduced in Osaka.

> This specification is based on and extends [Engine API - Prague](./prague.md) specification.

Reviewer 6: When new versions of methods are introduced there is usually a statement on how to limit the scope of the previous versions. I think this doc should at least add modification to `engine_getPayloadV4` while `engine_getBlobsV2` is a bit more tricky in that regard, see https://github.com/ethereum/execution-apis/blob/main/src/engine/prague.md#update-the-methods-of-previous-forks.

Author: Great callout, added `update-the-methods-of-previous-forks` section for `engine_getblobsv1`.

As for getPayloadV4, previous section already specifies the validation should reject anything that's older than Prague, so it covers Osaka.

Reviewer 6: > As for getPayloadV4, previous section already specifies the validation should reject anything that's older than Prague, so it covers Osaka.

`getPayloadV4` is introduced in Prague and it should also be mentioned that it must not be used for Osaka payloads as it is usually done

Author: right! got mixed up with v3, updated

- [Structures](#structures)
  - [BlobsBundleV2](#blobsbundlev2)
  - [BlobAndProofV2](#blobandproofv2)
- [Methods](#methods)
  - [engine_getPayloadV5](#engine_getpayloadv5)
    - [Request](#request)
    - [Response](#response)
    - [Specification](#specification)
  - [engine_getBlobsV2](#engine_getblobsv2)
    - [Request](#request-1)
    - [Response](#response-1)
    - [Specification](#specification-1)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Structures

### BlobsBundleV2

The fields are encoded as follows:

- `commitments`: `Array of DATA` - Array of `KZGCommitment` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).
- `proofs`: `Array of DATA` - Array of `KZGProof` (48 bytes each, type defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), semantics defined in [EIP-7594](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7594.md)).
- `blobs`: `Array of DATA` - Array of blobs, each blob is `FIELD_ELEMENTS_PER_BLOB * BYTES_PER_FIELD_ELEMENT = 4096 * 32 = 131072` bytes (`DATA`) representing a SSZ-encoded `Blob` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844)

> `blobs` and `commitments` arrays **MUST** be of same length and `proofs` **MUST** contain exactly `CELLS_PER_EXT_BLOB` * `len(blobs)` cell proofs.

Reviewer 7: It seems like this is the only way `BlobsBundleV2` differs from https://github.com/ethereum/execution-apis/blob/v1.0.0-beta.4/src/engine/cancun.md#blobsbundlev1?

### BlobAndProofV2

The fields are encoded as follows:

- `blob`: `DATA` - `FIELD_ELEMENTS_PER_BLOB * BYTES_PER_FIELD_ELEMENT = 4096 * 32 = 131072` bytes (`DATA`) representing a SSZ-encoded `Blob` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844).
- `proofs`: `Array of DATA` - Array of `KZGProof` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).

`proofs` **MUST** contain exactly `CELLS_PER_EXT_BLOB` cell proofs.

## Methods

> ### engine_getPayloadV5

Reviewer 8: How much effort is it to make these endpoints return in SSZ encoding? I'd imagine JSON serialisation and deserialisation would take much longer with much higher blob count.

Given that engine has SSZ encoding capability (blob is already SSZ encoded), is it possible to support SSZ encoding the entire endpoint responses?

Reviewer 8: Oh never mind - i think SSZ encoding the blobs already provide most of the savings on blobs.

Reviewer 8: Peter mentioned it before here:
https://x.com/peter_szilagyi/status/1841424117733478650

We should seriously consider doing this to support high blob count.

Reviewer 9: > We should seriously consider doing this to support high blob count.

the way the current engine-api is designed makes it almost impossible to add ssz as a wire format, yesterday the geth team mentioned they are working on a rest api following the beacon-api design, I think this is the right step forward, not only for blobs but also if we wanna bump up the gas limit

Reviewer 10: EIP-7919 would solve this as a side effect... it essentially makes everything SSZ (except the execution state).

This method is updated in a backward incompatible way. Instead of returning `BlobBundleV1`, it returns `BlobsBundleV2`.

#### Request

> * method: `engine_getPayloadV5`

Reviewer 6: This method should also be added to the OpenRPC schema

* params:
  1. `payloadId`: `DATA`, 8 Bytes - Identifier of the payload build process
* timeout: 1s

#### Response

* result: `object`
  - `executionPayload`: [`ExecutionPayloadV3`](./cancun.md#executionpayloadv3)
  - `blockValue` : `QUANTITY`, 256 Bits - The expected value to be received by the `feeRecipient` in wei
  - `blobsBundle`: [`BlobsBundleV2`](#BlobsBundleV2) - Bundle with data corresponding to blob transactions included into `executionPayload`
  - `shouldOverrideBuilder` : `BOOLEAN` - Suggestion from the execution layer to use this `executionPayload` instead of an externally provided one
  - `executionRequests`: `Array of DATA` - Execution layer triggered requests obtained from the `executionPayload` transaction execution.
* error: code and message set in case an exception happens while getting the payload.

#### Specification

This method follows the same specification as [`engine_getPayloadV4`](./prague.md#engine_getpayloadv4) with changes of the following:

> 1. The call **MUST** return `BlobsBundleV2` with empty `blobs`, `commitments` and `proofs` if the payload doesn't contain any blob transactions.

Reviewer 6: I guess this new method should return error when it is being called to return the payload pre-Osaka, see point (1) in [`engine_getPayloadV4`](https://github.com/ethereum/execution-apis/blob/main/src/engine/prague.md#engine_getpayloadv4) spec.

2. The call **MUST** return `blobs` and `proofs` that match the `commitments` list, i.e. 
   1. `assert len(blobsBundle.commitments) == len(blobsBundle.blobs)` and
   2. `assert len(blobsBundle.proofs) == len(blobsBundle.blobs) * CELLS_PER_EXT_BLOB` and
   3. `assert verify_cell_kzg_proof_batch(commitments, cell_indices, cells, blobsBundle.proofs)` (see [EIP-7594 consensus-specs](https://github.com/ethereum/consensus-specs/blob/36d80adb44c21c66379c6207a9578f9b1dcc8a2d/specs/fulu/polynomial-commitments-sampling.md#verify_cell_kzg_proof_batch))
      1. `commitments` should list each commitment `CELLS_PER_EXT_BLOB` times, repeating it for every cell. In python, `[blobsBundle.commitments[i] for i in range(len(blobsBundle.blobs)) for _ in range(CELLS_PER_EXT_BLOB)]`
      2. `cell_indices` should be `[0, ..., CELLS_PER_EXT_BLOB, 0, ..., CELLS_PER_EXT_BLOB, ...]`. In python, `list(range(CELLS_PER_EXT_BLOB)) * len(blobsBundle.blobs)`

>       3. `cells` is the list of cells for an extended blob. In python, `[compute_cells(blob) for blob in blobsBundle.blobs]` (see [compute_cells](https://github.com/ethereum/consensus-specs/blob/v1.5.0-beta.3/specs/fulu/polynomial-commitments-sampling.md#compute_cells) in consensus-specs)

Reviewer 4: Just realized that this isn't flattened. This should be a list of flattened cells, not a list of lists.

Author: Great catch, updated

      4. All of the inputs to `verify_cell_kzg_proof_batch` have the same length, `CELLS_PER_EXT_BLOB * len(blobsBundle.blobs)`

### engine_getBlobsV2

Consensus layer clients **MAY** use this method to fetch blobs from the execution layer blob pool.

#### Request

* method: `engine_getBlobsV2`
* params:

>   1. `Array of DATA`, 32 Bytes - Array of blob versioned hashes.

Reviewer 5: ```suggestion
  1. `Array of DATA`, 32 Bytes - Array of blob versioned hashes or `null` if ALL blobs not available with EL
```

* timeout: 1s

#### Response

> * result: `Array of BlobAndProofV2` - Array of [`BlobAndProofV2`](#BlobAndProofV2).

Reviewer 11: should it be `if ANY`? because partial response is not useful

> * result: `Array of BlobAndProofV2` - Array of [`BlobAndProofV2`](#BlobAndProofV2).

Reviewer 12: Some consensus around returning null: https://discord.com/channels/595666850260713488/1252403418941624532/1359823053269373099

Should also update `2.` in the Specification section below to indicate the response must be null

> * result: `Array of BlobAndProofV2` - Array of [`BlobAndProofV2`](#BlobAndProofV2).

Author: Updated to reflect the consensus

> * result: `Array of BlobAndProofV2` - Array of [`BlobAndProofV2`](#BlobAndProofV2), items of which may be `null`.

Reviewer 13: ```suggestion
* result: `Array of BlobAndProofV2` - Array of [`BlobAndProofV2`](#BlobAndProofV2).
```

Author: done

* error: code and message set in case an error occurs during processing of the request.

#### Specification

> Refer to the specification for [`engine_getBlobsV1`](./cancun.md#engine_getblobsv1) with changes of the following:

Reviewer 6: Are there actually any changes to the method behaviour? Afaics, this spec repeats the one for Cancun

Reviewer 4: Right now, the only difference is in first point, specifically stating that the response is an array of `BlobAndProofV2` objects now. But @flcl42 has a suggestion with more changes (which we should adopt):

* https://github.com/ethereum/execution-apis/pull/630/files#r2033047283

Reviewer: yeah we decided in the latest call to avoid partial responses, which are not useful to the CL, and instead do empty responses when not all blobs can be provided

> 1. Given an array of blob versioned hashes client software **MUST** respond with an array of `BlobAndProofV2` objects with matching versioned hashes, respecting the order of versioned hashes in the input array.

Reviewer 6: Should there be any information about how old transactions are handled by this method?

Author: It is specified here: https://github.com/ethereum/execution-apis/pull/630/files#diff-3a9e53de4a77d1ceec28870b24b0ff6e8f89437ae2d439aab70a75990060736dR35-R40

The idea is that old txs should've already been converted to new format or dropped already (which won't be included in the block)

Author: good call, updated point #2 to specify behavior for that situation

2. Client software **MUST** place responses in the order given in the request, using `null` for any missing blobs. For instance, if the request is `[A_versioned_hash, B_versioned_hash, C_versioned_hash]` and client software has data for blobs `A` and `C`, but doesn't have data for `B`, the response **MUST** be `[A, null, C]`.

> 3. Client software **MUST** support request sizes of at least 128 blob versioned hashes. The client **MUST** return `-38004: Too large request` error if the number of requested blobs is too large.

Reviewer 4: Should this limit be increased for future proofing? Maybe 256 or 512?

> 3. Client software **MUST** support request sizes of at least 128 blob versioned hashes. The client **MUST** return `-38004: Too large request` error if the number of requested blobs is too large.

Reviewer 4: Should this limit be increased for future proofing? Maybe 256 or 512?

> 2. Client software **MUST** place responses in the order given in the request, using `null` for any missing blobs. For instance, if the request is `[A_versioned_hash, B_versioned_hash, C_versioned_hash]` and client software has data for blobs `A` and `C`, but doesn't have data for `B`, the response **MUST** be `[A, null, C]`.
> 3. Client software **MUST** support request sizes of at least 128 blob versioned hashes. The client **MUST** return `-38004: Too large request` error if the number of requested blobs is too large.
> 4. Client software **MAY** return an array of all `null` entries if syncing or otherwise unable to serve blob pool data.
> 5. Callers **MUST** consider that execution layer clients may prune old blobs from their pool, and will respond with `null` if a blob has been pruned.

Reviewer 13: ```suggestion
2. Client software **MUST** return an empty array in case of any missing blobs. For instance, if the request is `[A_versioned_hash, B_versioned_hash, C_versioned_hash]` and client software has data for blobs `A` and `C`, but doesn't have data for `B`, the response **MUST** be `[]`.
3. Client software **MUST** support request sizes of at least 128 blob versioned hashes. The client **MUST** return `-38004: Too large request` error if the number of requested blobs is too large.
4. Client software **MAY** return an empty array if syncing or otherwise unable to serve blob pool data.
5. Callers **MUST** consider that execution layer clients may prune old blobs from their pool, and will respond with an empty array if a blob has been pruned.
```

Author: done

# Engine API -- Osaka

Engine API changes introduced in Osaka.

This specification is based on and extends [Engine API - Prague](./prague.md) specification.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Structures](#structures)
  - [ExecutionPayloadV4](#executionpayloadv4)

>   - [BlobsBundleV1](#blobsbundlev1)
>   - [BlobAndProofV1](#blobandproofv1)

Reviewer 14: ```suggestion
  - [BlobsBundleV2](#blobsbundlev2)
  - [BlobAndProofV2](#blobandproofv2)
```

Author: This is where the latest change is, the idea is not to bump the version of BlobsBundle and BlobAndProof message, just directly adding a new field to it, it could help simplify implementation quite a bit, and also more adhering to idiomatic protobuf conventions

Reviewer 14: > This is where the latest change is, the idea is not to bump the version of BlobsBundle and BlobAndProof message, just directly adding a new field to it, it could help simplify implementation quite a bit, and also more adhering to idiomatic protobuf conventions.

I see, it's not clear to me that this makes anything simpler, it seems like bumping a version is trivial, do you mind expanding on why this would be simpler? Is there an implementation that's made complex as a result of this?

Reviewer 14: ah I see, this suggests not even adding a new method

> engine_getPayloadV4

This might be fine then

Reviewer 5: to be honest i don't see much gain on both EL and CL side by avoiding version bump, infact it add for easy validation and typestrictness with fork rather than now having both `proof` and `cellProofs` as optionals

i would recommend the version bump

Reviewer 8: Agree with @g11tech. I think version bump is a bit cleaner than having an extra field that's always going to be empty from Fulu.

Reviewer 8: How do people feel about keeping the same field name (`proofs`), but bumping the API version? 

From the CL side, I feel like we could potentially get the best of both worlds with this - easy validations, while keeping the structure unchanged. Separately, we also have a `BlobsBundle` in the Builder API, which is currently identical to this, and we could keep the SSZ encoding unchanged if the field remains the same (max length of list does not affect SSZ encoding).

Reviewer 11: IMO, I prefer bump the version. instead of modify and/or add fields to existing types.

Reviewer 15: strong preference on bumping the API versions

Reviewer 14: After thinking about it, also much prefer the bump, since semantics are changing.

- `parentHash`: `DATA`, 32 Bytes
- `feeRecipient`:  `DATA`, 20 Bytes
- `stateRoot`: `DATA`, 32 Bytes
- `receiptsRoot`: `DATA`, 32 Bytes
- `logsBloom`: `DATA`, 256 Bytes
- `prevRandao`: `DATA`, 32 Bytes
- `blockNumber`: `QUANTITY`, 64 Bits
- `gasLimit`: `QUANTITY`, 64 Bits
- `gasUsed`: `QUANTITY`, 64 Bits
- `timestamp`: `QUANTITY`, 64 Bits
- `extraData`: `DATA`, 0 to 32 Bytes
- `baseFeePerGas`: `QUANTITY`, 256 Bits
- `blockHash`: `DATA`, 32 Bytes
- `transactions`: `Array of DATA` - Array of transaction objects, each object is a byte list (`DATA`) representing `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)
- `withdrawals`: `Array of WithdrawalV1` - Array of withdrawals, each object is an `OBJECT` containing the fields of a `WithdrawalV1` structure.
- `blobGasUsed`: `QUANTITY`, 64 Bits
- `excessBlobGas`: `QUANTITY`, 64 Bits
- `proofVersion`: `QUANTITY`, 8 Bits

Currently `proofVersion` supports two types:

- `0`: original type, blob proofs
- `1`: new type after EIP-7594, cell proofs

> ### BlobsBundleV1

Reviewer 14: ```suggestion
### BlobsBundleV2
```

This structure has the syntax of [`BlobsBundleV1`](./cancun.md#blobsbundlev1) and appends the new fields: `cellProofs`.

- `commitments`: `Array of DATA` - Array of `KZGCommitment` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).
- `proofs`: `Array of DATA` - Array of `KZGProof` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), 48 bytes each (`DATA`).
- `blobs`: `Array of DATA` - Array of blobs, each blob is `FIELD_ELEMENTS_PER_BLOB * BYTES_PER_FIELD_ELEMENT = 4096 * 32 = 131072` bytes (`DATA`), representing a SSZ-encoded `Blob` as defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844)
- `cellProofs`: `Array of DATA` - Array of `KZGProof` (48 bytes each, type defined in [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844), semantics defined in [EIP-7594](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7594.md)).

`blobs` and `commitments` arrays **MUST** be of same length, `cellProofs` contains exactly `CELLS_PER_EXT_BLOB` * `len(blobs)` cell proofs.
Before Osaka fork, `cellProofs` should be empty, and after `proof` should be empty.

> ### BlobAndProofV1

Reviewer 14: ```suggestion
### BlobAndProofV2
```

1. Given an array of blob versioned hashes client software **MUST** respond with an array of `BlobAndProofV2` objects with matching versioned hashes, respecting the order of versioned hashes in the input array.
2. Client software **MUST** return `null` in case of any missing blobs. For instance, if the request is `[A_versioned_hash, B_versioned_hash, C_versioned_hash]` and client software has data for blobs `A` and `C`, but doesn't have data for `B`, the response **MUST** be `null`.
3. Client software **MUST** support request sizes of at least 128 blob versioned hashes. The client **MUST** return `-38004: Too large request` error if the number of requested blobs is too large.
4. Client software **MUST** return `null` if syncing or otherwise unable to serve blob pool data.
5. Callers **MUST** consider that execution layer clients may prune old blobs from their pool, and will respond with `null` if a blob has been pruned.

### Update the methods of previous forks

This section defines how Osaka payload should be handled by the ['Cancun API'](./cancun.md).

For the following methods:

- [`engine_getBlobsV1`](./cancun.md#engine_getblobsv1)

a validation **MUST** be added:

> 1. Client software **MUST** return `-38005: Unsupported fork` error if the `timestamp` of payload greater or equal to the Osaka activation timestamp.

Reviewer 6: What payload it should get the timestamp from?

Author: updated the wording to `if the Osaka fork has been activated`
